server/src/loggers.rs

```diff
@@ -49,7 +64,143 @@ pub fn init(log_path: &Path, log_name: String, std_colors: bool) -> Result<Dispa
+        })
+        .chain(Box::new(pdl) as Box<dyn Log>)
+}
+
+struct PDLogger {
+    key: String,
+    handle: Handle,
+    build: String,
+}
+
+impl Log for PDLogger {
+    fn enabled(&self, metadata: &Metadata) -> bool {
+        metadata.level() <= Level::Error
+    }
+
+    fn log(&self, record: &Record) {
+        if self.enabled(record.metadata()) {
+            notify(
+                &self.key,
+                &self.handle,
+                Event::AlertTrigger(AlertTrigger {
+                    payload: AlertTriggerPayload {
+                        severity: level_to_severity(record.level()),
+                        summary: record.args().to_string(),
+                        source: "localhost".to_string(), // TODO: Hostname
+                        timestamp: Some(SystemTime::now().into()),
+                        component: None,
+                        group: None,
+                        class: None,
+                        custom_details: Some(LogDetails {
+                            data: record.args().to_string(),
+                            logger: record.target().to_string(),
+                            file: record.file().map(|c| c.to_string()),
+                            line: record.line().map(|c| c.to_string()),
+                            build: self.build.to_string(),
+                        }),
+                    },
+                    dedup_key: Some(dedup_key(record, self.build.to_string())),
+                    images: None,
+                    links: None,
+                    client: None,
+                    client_url: None,
+                }),
+            );
+        }
+    }
+
+    fn flush(&self) {}
+}
+
+fn level_to_severity(level: Level) -> Severity {
+    match level {
+        Level::Error => Severity::Error,
+        Level::Warn => Severity::Info,
+        Level::Info => Severity::Info,
+        Level::Debug => Severity::Info,
+        Level::Trace => Severity::Info,
+    }
+}
+
+fn dedup_key(record: &Record, build: String) -> String {
+    match (record.file(), record.line()) {
+        (Some(file), Some(line)) => {
+            format!("{}-{}#{}", build, file, line)
+        }
+        (_, _) => build
+            .chars()
+            .chain(record.target().chars().take(30))
+            .chain("::".chars())
+            .chain(record.args().to_string().chars())
+            .take(255)
+            .collect(),
+    }
+}
+
+#[derive(Serialize)]
+struct LogDetails<T: serde::Serialize> {
+    data: T,
+    logger: String,
+    file: Option<String>,
+    line: Option<String>,
+    build: String,
+}
+
+#[derive(Serialize)]
+struct ChangeDetail {
+    build: String,
+}
+
+fn notify<T: serde::Serialize + std::marker::Send + std::marker::Sync + 'static>(
+    api_key: &String,
+    handle: &Handle,
+    event: Event<T>,
+) {
+    let events = EventsV2::new(api_key.to_string(), Some("lockbook-server".to_string())).unwrap();
+
+    futures::executor::block_on(async {
```

Comment:
* first of all can we use tokio's block on instead of futures? no need for an extra dependency
* second of all, are you sure this is okay? I expect this to crash when called from an async function. sync -> async -> sync -> async is not allowed because creating an sync runtime is a blocking operation which is not ok in async functions
* shouldn't this be an async function anyway? I know this makes life hard because [traits don't support async functions yet last I checked](http://smallcultfollowing.com/babysteps/blog/2019/10/26/async-fn-in-traits-are-hard/). I struggled making our dependency injection pattern work using that crate, which is why server doesn't have dependency injection right now

Reply:
I don't think this is creating an async runtime but rather giving async work back to tokio via `handle`

Earlier with the sync version of this library I was experiencing those crashes since it was trying to start an async runtime before running the same async code

Yeah maybe it being async _does_ make sense but that would be violating all of the logging framework stuff. However, if we could have a "fire-and-forget" style piece of async code, that would be ideal since that's kinda what we want out of PagerDuty

https://stackoverflow.com/a/62536772